### ISSUE TITLE
Assembly not marked and asm editor doesn't scroll to selected source line

### ISSUE THREAD

**Issue by luni64** [MRobertEvers/vscode-compiler-explorer#20]
I'm using your extension in an embedded c++ build system using a local installation of CE (see here for a short test video https://youtu.be/6S5EomO1Dy4). 

Everything works quite well but most of the time your extension doesn't mark the corresponding assembly if I select a source line.  It sometimes works but not always. I wasn't able to detect any pattern when it works and when not. 

Here the generated settings.json: 
```json
{
  "task.autoDetect": "off",
  "compiler-explorer.url": "http://localhost:10240",
  "compiler-explorer.compiler": "g54",
  "compiler-explorer.options": [
    "-I'C:/Users/lutz/source/teensy/TeensySupport/TeensyStep/overrideSpeed/src'",
    "-I'C:/Users/lutz/source/teensy/TeensySupport/TeensyStep/overrideSpeed/cores/teensy3'",
    "-I'C:/Users/lutz/source/teensy/TeensySupport/TeensyStep/overrideSpeed/lib/ResponsiveAnalogRead-1.2.1/src'",
    "-I'C:/Users/lutz/source/teensy/TeensySupport/TeensyStep/overrideSpeed/lib/TeensyStep/src'",
    "-D__MK66FX1M0__ -DTEENSYDUINO=156",
    "-DF_CPU=180000000",
    "-DUSB_SERIAL",
    "-DLAYOUT_US_ENGLISH",
    "-DARDUINO_TEENSY36",
    "-DARDUINO=10815",
    "-fno-exceptions -fpermissive -felide-constructors -std=gnu++14 -Wno-error=narrowing -fno-rtti",
    "-mthumb -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -fsingle-precision-constant",
    "-O2",
    "-g -Wall -ffunction-sections -fdata-sections -nostdlib -mno-unaligned-access",
  ]
}
```
Am I missing some configuration option?

**Comment by MRobertEvers** [comment]
I'm not familiar with that compiler, but here are a few thoughts. 

1. With optimizations enabled, sometimes source lines are omitted entirely. 
2. Compiler explorer uses the output from the selected compiler to determine lines. Usually the compiler has a switch to produce assembly with line numbers.

See this example from here for MSVC, https://stackoverflow.com/questions/1289881/using-gcc-to-produce-readable-assembly.

Can you look at the disassembled output for your project with the line numbers switch? If lines are optimized out, Compiler Explorer won't know what lines to highlight.

For example, GCC's objdump has the `--line-numbers` switch.

**Comment by MRobertEvers** [comment]
Here is a quick example with clang.

```c
// main.c
#include <stdio.h>
int main() 
{
    printf("Hello");
    return 0;
}
```

Compile with `clang -c -g main.c`. The disassemble with `llvm-objdump -Sl main.o`.

Produces

```
; main():
; /Users/matthewevers/Documents/git_repos/objdump/main.c:4
; {
       0: ff 83 00 d1   sub     sp, sp, #32
       4: fd 7b 01 a9   stp     x29, x30, [sp, #16]
       8: fd 43 00 91   add     x29, sp, #16
       c: 08 00 80 52   mov     w8, #0
      10: e8 0b 00 b9   str     w8, [sp, #8]
      14: bf c3 1f b8   stur    wzr, [x29, #-4]
      18: 00 00 00 90   adrp    x0, 0x0 <ltmp0+0x18>
      1c: 00 00 00 91   add     x0, x0, #0
; /Users/matthewevers/Documents/git_repos/objdump/main.c:5
;     printf("Hello");
      20: 00 00 00 94   bl      0x20 <ltmp0+0x20>
      24: e0 0b 40 b9   ldr     w0, [sp, #8]
; /Users/matthewevers/Documents/git_repos/objdump/main.c:6
;     return 0;
      28: fd 7b 41 a9   ldp     x29, x30, [sp, #16]
      2c: ff 83 00 91   add     sp, sp, #32
      30: c0 03 5f d6   ret
```

Notice the line numbers are included (`-l` is short for `--line-numbers`). Can you produce something similar for your compiler? You can check if a line number is included for the lines you want to highlight.

**Comment by luni64** [comment]
Thanks for your answer. I did an objdump of the following source code. (GCC 5.4, ARM-NONE-EABI)

**SOURCE**
```c++
#include "Arduino.h"

void setup()
{
  pinMode(LED_BUILTIN,OUTPUT);

  Serial.print("test");
}

void loop()
{
  digitalToggleFast(LED_BUILTIN);
  digitalToggleFast(LED_BUILTIN);
  digitalToggleFast(LED_BUILTIN);
  digitalToggleFast(LED_BUILTIN);
}
```
It generates the following asm: 

**ASM**
```
setup:
        push    {r3, lr}
        movs    r1, #1
        movs    r0, #13
        bl      pinMode
        movs    r1, #4
        ldr     r0, .L3
        pop     {r3, lr}
        b       usb_serial_write
.L3:
        .word   .LC0
loop:
        ldr     r3, .L6
        movs    r2, #8
        str     r2, [r3, #140]
        str     r2, [r3, #140]
        str     r2, [r3, #140]
        str     r2, [r3, #140]
        bx      lr
.L6:
        .word   1107312640
.LC0:
        .ascii  "test\000"
```
- Selecting the opening brace after `setup()` highlights the `push {r3, lr}`
- Selecting the pinMode line highlights the corresponding lines in the CE window
- Selecting any other line does nothing.

**OBJDUMP**
objdump -d -S --demangle --line-numbers main.cpp.o gives the following output: 

```
arm-none-eabi-objdump -d -S --demangle --line-numbers  main.cpp.o  

main.cpp.o:     file format elf32-littlearm

Disassembly of section .text.setup:

00000000 <setup>:
setup():
C:\Users\lutz\source\teensy\01_Scratch\tst4/src/main.cpp:4
#include "Arduino.h"

void setup()
{
   0:   b508            push    {r3, lr}
C:\Users\lutz\source\teensy\01_Scratch\tst4/src/main.cpp:5
  pinMode(LED_BUILTIN,OUTPUT);
   2:   2101            movs    r1, #1
   4:   200d            movs    r0, #13
   6:   f7ff fffe       bl      0 <pinMode>
_ZN16usb_serial_class5writeEPKhj():
C:\toolchain\Arduino\arduino-1.8.15\hardware\teensy\avr\cores\teensy4/usb_serial.h:96
        virtual int read() { return usb_serial_getchar(); }
        virtual int peek() { return usb_serial_peekchar(); }
        virtual void flush() { usb_serial_flush_output(); }  // TODO: actually wait for data to leave USB...
        virtual void clear(void) { usb_serial_flush_input(); }
        virtual size_t write(uint8_t c) { return usb_serial_putchar(c); }
        virtual size_t write(const uint8_t *buffer, size_t size) { return usb_serial_write(buffer, size); }
   a:   2104            movs    r1, #4
   c:   4802            ldr     r0, [pc, #8]    ; (18 <setup+0x18>)
setup():
C:\Users\lutz\source\teensy\01_Scratch\tst4/src/main.cpp:8

  Serial.print("test");
}
   e:   e8bd 4008       ldmia.w sp!, {r3, lr}
_ZN16usb_serial_class5writeEPKhj():
C:\toolchain\Arduino\arduino-1.8.15\hardware\teensy\avr\cores\teensy4/usb_serial.h:96
  12:   f7ff bffe       b.w     0 <usb_serial_write>
  16:   bf00            nop
  18:   00000000        .word   0x00000000

Disassembly of section .text.loop:

00000000 <loop>:
loop():
C:\toolchain\Arduino\arduino-1.8.15\hardware\teensy\avr\cores\teensy4/core_pins.h:2189
                } else if (pin == 11) {
                        CORE_PIN11_PORTTOGGLE = CORE_PIN11_BITMASK;
                } else if (pin == 12) {
                        CORE_PIN12_PORTTOGGLE = CORE_PIN12_BITMASK;
                } else if (pin == 13) {
                        CORE_PIN13_PORTTOGGLE = CORE_PIN13_BITMASK;
   0:   4b05            ldr     r3, [pc, #20]   ; (18 <loop+0x18>)
   2:   2208            movs    r2, #8
   4:   f8c3 208c       str.w   r2, [r3, #140]  ; 0x8c
   8:   f8c3 208c       str.w   r2, [r3, #140]  ; 0x8c
   c:   f8c3 208c       str.w   r2, [r3, #140]  ; 0x8c
  10:   f8c3 208c       str.w   r2, [r3, #140]  ; 0x8c
digitalToggleFast():
C:\toolchain\Arduino\arduino-1.8.15\hardware\teensy\avr\cores\teensy4/core_pins.h:2189
  14:   4770            bx      lr
  16:   bf00            nop
  18:   42004000        .word   0x42004000
```

**Comment by MRobertEvers** [comment]
It looks like GCC is inlining a lot of your code. If the compiler inlines code from other source files, then compiler explorer doesn't know what lines to highlight. 

Look at `Serial.print("test");`, the compiler says that this line is inlined from `C:\toolchain\Arduino\arduino-1.8.15\hardware\teensy\avr\cores\teensy4/usb_serial.h:96`. 

`loop()` looks like it is almost entirely inlined. So I'm guessing you're not seeing anything in that function highlighted.

Additionally, if the Teensy libraries use macros, the compiler will also have problems identifying which source line to associate with the assembly.

**Comment by luni64** [comment]
Thanks for analyzing this. Yes, this is highly optimized embedded code, a lot of stuff will be inlined. I think I understand the underlying problem now. I also tried the same code with the web GUI and get the same result which is clear now. 

Anyway, this is a great extension and very useful for optimizing stuff. Thanks for sharing it. I implemented CE support (auto generation of CE entries in settings.json) in my VisualTeensy build system. Lets see what the users say...
